### PR TITLE
[Terrarium - Providers]: Implementation of Storage Feature - 1

### PR DESCRIPTION
**Pull request by anjanikshree12**
### This PR includes:
* Protobuff definitions for storage service having functionalities such as download provider binary, shasum and signature files which is a required metadata that a particular version and platform of a provider requires.

* gRPC code generated by protoc compiler tool from the protobuff definitions.

## Description:
* The plan is to extend the rest protocol implementation of providers v1, to design 3 APIs that could be used to download the required files/provider binaries as per the provider registry protocol from the storage system.
```
https://<host-name>/providers/v1/{organization_name}/{name}/{version}/{os}/{arch}/terraform-provider-{name}_{version}_{os}_{arch}.zip
https://<host-name>/providers/v1/{organization_name}/{name}/{version}/terraform-provider-{name}_{version}_SHA256SUMS
https://<host-name>/providers/v1/{organization_name}/{name}/{version}/terraform-provider-{name}_{version}_SHA256SUMS.sig
```
* The files needed as per the provider registry protocol for download through the URL is provider binary archive, shasum file & shasum signature file.
```
"download_url": "https://<host-name>/providers/v1/{organization_name}/{name}/{version}/{os}/{arch}/terraform-provider-{name}_{version}_{os}_{arch}.zip",
"shasums_url": "https://<host-name>/providers/v1/{organization_name}/{name}/{version}/terraform-provider-{name}_{version}_SHA256SUMS",
"shasums_signature_url": "https://<host-name>/providers/v1/{organization_name}/{name}/{version}/terraform-provider-{name}_{version}_SHA256SUMS.sig"
```
## S3 Directory tree structure to be followed.
```
--organisation1
     |---- provider1
              |------ version
                         |----- terraform-provider-<provider-name>_<version>_SHA256SUMS
                         |----- terraform-provider-<provider-name>_<version>_SHA256SUMS.sig
                         |----- terraform-provider-<provider-name>_<version>_<os>_<arch>.zip
                         |----- terraform-provider-<provider-name>_<version>_<os>_<arch>.zip
                         continued ...
              |------ version2
                         |----- terraform-provider-<provider-name>_<version>_SHA256SUMS
                         |----- terraform-provider-<provider-name>_<version>_SHA256SUMS.sig
                         |----- terraform-provider-<provider-name>_<version>_<os>_<arch>.zip
                         |----- terraform-provider-<provider-name>_<version>_<os>_<arch>.zip
                         continued ...
     |---- provider2
              |------ version
                         |----- terraform-provider-<provider-name>_<version>_SHA256SUMS
                         |----- terraform-provider-<provider-name>_<version>_SHA256SUMS.sig
                         |----- terraform-provider-<provider-name>_<version>_<os>_<arch>.zip
                         |----- terraform-provider-<provider-name>_<version>_<os>_<arch>.zip
                         continued ...
              |------ version2
                         |----- terraform-provider-<provider-name>_<version>_SHA256SUMS
                         |----- terraform-provider-<provider-name>_<version>_SHA256SUMS.sig
                         |----- terraform-provider-<provider-name>_<version>_<os>_<arch>.zip
                         |----- terraform-provider-<provider-name>_<version>_<os>_<arch>.zip
                         continued ...
```
Note: Shasum file and signature is common for each version, so getting stored at version level.

### CODE REVIEW

**Review by odelanno — CHANGES_REQUESTED** [review]
gen_proto.sh needs to be upgraded to include the storage protocol buffer file so that everything is re generated correctly.

**Review by odelanno — APPROVED** [review]

File: pb/terrarium/provider/provider.proto [hunk 1]
```diff
@@ -8,6 +8,12 @@ service ProviderPublisher {
   rpc EndProvider(EndProviderRequest) returns (Response) {}
 }
 
+service ProviderConsumer {
```
Comment (adcharre) [inline pb/terrarium/provider/provider.proto:11]: We don't need this service only the internal ones as the ability to download the various files is via a rest API like for modules. The Rest API handler will directly use the internal service.

File: pb/terrarium/provider/provider.proto [hunk 2]
```diff
@@ -72,4 +78,27 @@ message EndProviderRequest {
...
+  bytes zip_data_chunk = 1;
+}
+
+message DownloadShasumRequest {
+  Provider provider = 1;
+}
+
+message DownloadShasumResponse {
+  bytes shasum_data_chunk = 1;
+}
```
Comment (adcharre) [inline pb/terrarium/provider/provider.proto:104]: Move these to pb/terrarium/provider/services/storage.proto.

Comment (anjanikshree12) [reply pb/terrarium/provider/provider.proto]: Moved.

**Review by adcharre — DISMISSED** [review]

**Review by adcharre — APPROVED** [review]

**Pull request merged**